File: java-common/src/main/java/com/typesafe/conductr/lib/java/ConnectionHandler.java
package com.typesafe.conductr.lib.java;

import com.typesafe.conductr.lib.HttpPayload;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.net.HttpURLConnection;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;
import java.util.concurrent.Executor;
import java.util.function.BiFunction;
import java.util.stream.Collectors;

import static com.typesafe.conductr.lib.java.ManagedBlocker.*;

/**
 * Connection handlers provide the means to establish a connection, issue a request and then finalize
 * the connection
 */
public class ConnectionHandler {

    protected final static String USER_AGENT = "TypesafeConductRLib";

    /**
     * Make a request to a ConductR service given a payload. Returns a future of an option. If there is some response
     * then there will be a result, otherwise empty indicates that this program is not running in the context
     * of ConductR.
     */
    public static <T> CompletionStage<Optional<T>> withConnectedRequest(
            Optional<HttpPayload> payload, BiFunction<Integer, Map<String, Optional<String>>, Optional<T>> op, Executor executor) {
        return payload
            .map(p -> CompletableFuture.supplyAsync(() -> {
                try {
                    HttpURLConnection connection = (HttpURLConnection) p.getUrl().openConnection();
                    connection.setRequestMethod(p.getRequestMethod());
                    connection.setInstanceFollowRedirects(p.getFollowRedirects());
                    connection.setRequestProperty("User-Agent", USER_AGENT);
                    blocking(() -> {
                        try {
                            connection.connect();
                        } catch (IOException e) {
                            throw new UncheckedIOException(e);
                        }
                    });
                    Map<String, Optional<String>> headers =
                            connection.getHeaderFields()
                                    .entrySet().stream()
                                    .collect(Collectors.toMap(Map.Entry::getKey, e -> e.getValue().stream().reduce((a, b) -> b)));
                    return op.apply(connection.getResponseCode(), headers);

                } catch (IOException e) {
                    throw new UncheckedIOException("Connection failed for " + p, e);
                } catch (InterruptedException e) {
                    throw new UncheckedIOException("Interrupt", new IOException("Connection failed for " + p));
                }
            }, executor))
            .orElse(CompletableFuture.completedFuture(Optional.empty()));
    }
}

File: scala-common/src/main/scala/com/typesafe/conductr/lib/scala/ConnectionHandler.scala
package com.typesafe.conductr.lib.scala

import java.io.IOException
import java.net.HttpURLConnection
import com.typesafe.conductr.lib.HttpPayload
import scala.concurrent._
import scala.util.{ Failure, Success, Try }

object ConnectionContext {
  def apply(executionContext: ExecutionContext): ConnectionContext =
    new ConnectionContext()(executionContext)

  object Implicits {
    /**
     * An implicit global ConnectionContext.
     * Import global when you want to provide the global ConnectionContext implicitly.
     * This global ConnectionContext uses Scala's global execution context.
     */
    implicit val global = ConnectionContext(ExecutionContext.Implicits.global)
  }
}

/**
 * When performing pure Scala connections, this is the connection context to use. Pass in the
 * execution context to be used for blocking IO.
 */
class ConnectionContext()(implicit val executionContext: ExecutionContext) extends AbstractConnectionContext

/**
 * INTERNAL API
 * Handles the JDK HttpURLConnection requests and responses
 */
class ConnectionHandler extends AbstractConnectionHandler {

  override protected type CC = ConnectionContext

  /**
   * Make a request to a ConductR service given a payload. Returns a future of an option. If there is some response
   * then a Some() will convey the result, otherwise None indicates that this program is not running in the context
   * of ConductR.
   */
  override def withConnectedRequest[T](
    payload: Option[HttpPayload])(handler: (Int, Map[String, Option[String]]) => Option[T])(implicit cc: CC): Future[Option[T]] = {

    import cc.executionContext
    payload.fold[Future[Option[T]]](Future.successful(None)) { p =>
      Future {
        Try(p.getUrl.openConnection) match {
          case Success(connection: HttpURLConnection) =>
            connection.setRequestMethod(p.getRequestMethod)
            connection.setInstanceFollowRedirects(p.getFollowRedirects)
            connection.setRequestProperty("User-Agent", UserAgent)
            blocking {
              connection.connect()
              import scala.collection.JavaConverters._
              try {
                handler(
                  connection.getResponseCode,
                  connection.getHeaderFields.asScala.foldLeft(Map.empty[String, Option[String]]) {
                    case (m, (k, v)) => m.updated(k, v.asScala.lastOption)
                  })

              } finally {
                connection.disconnect()
              }
            }
          case Success(connection) =>
            throw new IOException(s"Unexpected type of connection $connection for $p")
          case Failure(e) =>
            throw new IOException(s"Connection failed for $p", e)
        }
      }
    }
  }

}
